src/Agent.Worker/StepsRunner.cs

```diff
@@ -205,5 +204,36 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 Trace.Info($"Current state: job state = '{jobContext.Result}', step failed = {stepFailed}, critical step failed = {criticalStepFailed}");
             }
         }
+
+        private bool ShouldRunStep(IStep step, bool stepFailed, bool criticalStepFailed)
+        {            
+            switch(step.RunMode)
+            {
+                // The server hasn't been upgraded yet. Fallback to old code
+                case RunMode.Undefined:
+                    // TODO: Run finally even if canceled?
+                    // Run if the step if Finally or no step has failed or the step is always run and no criticalstep has failed
+                    return step.Finally || !stepFailed || (!criticalStepFailed && step.AlwaysRun);
+                default:
+                    return step.Finally || ShouldRunStepBasedOnRunMode(step, stepFailed, criticalStepFailed);
+            }
+        }
+
+        private static bool ShouldRunStepBasedOnRunMode(IStep step, bool stepFailed, bool criticalStepFailed)
+        {
+            switch(step.RunMode)
+            {
+                case RunMode.Default:
+                    return !stepFailed;
+                case RunMode.Always:
+                    return !criticalStepFailed;
+                case RunMode.Rollback:
+                    return stepFailed && !criticalStepFailed;
+                case RunMode.Custom:
+                    return !stepFailed && TaskConditionsEvaluator.AreConditionsSatisfied(step.Conditions, step.ExecutionContext.Variables, step.ExecutionContext);
```

Comment:
it looks like this is wrong. it does not use the latest expanded variables. the agent initalizes the variables once, indirectly in jobrunner when `jobContext.InitializeJob` is called. then more variables are set. then almost immediately before `step.RunAsync` is called expansion is recalculated.

your conditional check is too early